include/onnxruntime/core/framework/data_types.h

```diff
@@ -207,19 +208,49 @@ class DataTypeImpl {
   static void RegisterDataType(MLDataType);
   static MLDataType GetDataType(const std::string&);
 
-  static const std::vector<MLDataType>& AllTensorTypes();
-  static const std::vector<MLDataType>& AllFixedSizeTensorTypes();
-  static const std::vector<MLDataType>& AllSequenceTensorTypes();
-  static const std::vector<MLDataType>& AllFixedSizeSequenceTensorTypes();
-  static const std::vector<MLDataType>& AllNumericTensorTypes();
-  static const std::vector<MLDataType>& AllIEEEFloatTensorTypes();
-  static const std::vector<MLDataType>& AllFixedSizeTensorExceptHalfTypes();
-  static const std::vector<MLDataType>& AllIEEEFloatTensorExceptHalfTypes();
-  static const std::vector<MLDataType>& AllTensorAndSequenceTensorTypes();
-  static const std::vector<MLDataType>& AllOptionalAndTensorAndSequenceTensorTypes();
-  static const std::vector<MLDataType>& AllFixedSizeTensorAndSequenceTensorTypes();
-  static const std::vector<MLDataType>& AllOptionalTypes();
-  static const std::vector<MLDataType>& AllTensorAndSequenceTensorAndOptionalTypes();
+  // IR4: includes all float types, includes float16, bfloat16
+  // IR9: includes float 8 types as well
+  static const std::vector<MLDataType>& AllTensorTypes();  // up to IR4 (no float 8), deprecated
+  static const std::vector<MLDataType>& AllTensorTypesIR4();
+  static const std::vector<MLDataType>& AllTensorTypesIR9();
```

Comment:
That was the previous names and there was a comment about following onnx names. I don't remember exactly where it was added in the PR. I prefer avoiding to remove `AllTensorTypes` to avoid breaking the C API. What about adding AllTensorTypesWithFloat8 to make it equal to AllTensorTypesIR9?

Reply:
I agree with Scott, IR4 and IR9 is not intuitive. 

I think having AllTensorTypes and AllTensorTypesWithFloat8 is sufficient. This way only the ops which support FP8 can change to AllTensorTypesWithFloat8. Same applies to all other cases below.

Reply:
What would break in the C API? I didn't see a public usage of AllTensorTypes there.

I agree having an unexpected meaning for 'all' is potentially confusing.
That 'IR4' or 'IR9' relates to an ONNX IR version is also potentially confusing to an ORT user. 
We also don't support complex64 and complex128 so the ORT 'all' in any case doesn't equate to all the types the ONNX spec has.

Given the extensive usage of AllTensorTypes in the kernel registrations of existing operators, I'd lean towards keeping that as-is, and adding AllTensorTypesNoFloat8/AllTensorTypesWithFloat8 as the preferred names. If we want to have a more explicit link with the ONNX IR versions maybe that can be an implementation detail like this:

```c++
private:
  static const std::vector<MLDataType>& AllTensorTypesOnnxIRv4(); 
  static const std::vector<MLDataType>& AllTensorTypesOnnxIRv9();  // added 8-bit float types
...
public:
  static const std::vector<MLDataType>& AllTensorTypes() { return AllTensorTypesOnnxIRv4(); } // for backwards compatibility
  static const std::vector<MLDataType>& AllTensorTypesNoFloat8() { return AllTensorTypesOnnxIRv4(); }
  static const std::vector<MLDataType>& AllTensorTypesWithFloat8() { return AllTensorTypesOnnxIRv9(); }
```

Ideally we'd replace the current usage of AllTensorTypes with AllTensorTypesNoFloat8 to be more explicit and if possible remove AllTensorTypes. And if that removal is possible, AllTensorTypesWithFloat8 can be renamed to AllTensorTypes.

Reply:
By C API, I meant `provider_interfaces.h`. I was not sure if this API was internal or public. Before I rename, I think we may assume there will be other float types. These definitions are mostly to register kernels. What would the rule in that case?

Reply:
I renamed IR4, IR9 into IRv4, IRv9.